Implement MapReduce mapper

sprzedwojski opened the pull request:
Closes #20 

The example used to showcase the MapReduce mapper is a wordcount program, which counts the number of occurrences of each word in an input text file.

[Example input Oozie workflow XML](https://github.com/GoogleCloudPlatform/cloud-composer/blob/map-reduce/oozie-to-airflow/examples/mapreduce/workflow.xml)

[Directory with all example files](https://github.com/GoogleCloudPlatform/cloud-composer/tree/map-reduce/oozie-to-airflow/examples/mapreduce)

[Example converted DAG code](https://gist.github.com/sprzedwojski/28b5580b4a7becb32cd4fdd9fcc9b884)

Example converted DAG graph:
![image](https://user-images.githubusercontent.com/4280682/56299276-ff8a7080-6133-11e9-81dc-4fed2509c026.png)

mik-laj commented:
Can you visualize and source codes of result files?
Similar to: https://github.com/GoogleCloudPlatform/cloud-composer/pull/61

mik-laj commented:
in the PR title if you use e.g. "Closes #20" it will close the issue automatically once this PR is merged.
https://help.github.com/en/articles/closing-issues-using-keywords

sprzedwojski commented:
I've added a commit in which there's a new functionality in `el_utils` to enable parsing a config line with a comma-separated list, e.g.
`hadoop_jars=1.jar,2.jar`
Also added a unit test for this.

Finally changed some field names related to File/Archive extractors which I forgot to update in their PR (already merged).

mik-laj commented on oozie-to-airflow/mappers/__init__.py:
```diff
@@ -20,8 +20,8 @@
     "dummy_mapper",
-    "file_archive_mappers",
     "kill_mapper",
+    "mapreduce_mapper",
```
Why does this file have content?  In our case, an empty file is sufficient

mik-laj replied:
In my opinion, this is not a good solution. This import should also work if we do not define any value.
I created an experiments. I deleted the contents of the init files....and everything is working properly.
https://github.com/GoogleCloudPlatform/cloud-composer/commits/cleanup-init

Property `__all__` in module is import, when we write libraries. This allows you to limit the available module classes on the outside.  We do not design module interfaces, so it does not matter to us.

potiuk replied:
OK. Great! It must have been some transient problem then while I was adding the static checkers. Why don't you make a PR removing it? I think it should be a separate PR to remove it and this one should go "as is"

mik-laj replied:
I did not want to create another place for discussion.  I will create a PR when we get a consensus. if I would not be right, then a separate PR creates noise.

potiuk replied:
You misunderstood my English Kamil :). 

I think we are in perfect understanding and consensus already but there is a (small) communication problem.

I like that we had discussion here. And I love you disagreed with me and presented good arguments. And I agreed with them and we reached consensus already. The "Why don't you ?" question was not to ask you "Why did not you create it in the first place?" it was  - please go ahead and create PR. 

That is a subtle difference in English when you use past vs. present tense which you apparently did not realise :).

potiuk replied:
I think you misunderstood my English Kamil :). 

I think we are in perfect understanding and consensus already but there is a (small) communication problem.

I like that we had discussion here. And I love you disagreed with me and presented good arguments. And I agreed with them and we reached consensus already. The "Why don't you ?" question was not to ask you "Why did not you create it in the first place?" it was  - please go ahead and create PR. 

That is a subtle difference in English when you use past vs. present tense which you apparently did not realise :).

mik-laj replied:
Yes. I was currently busy with other activities and I did not focus enough on your message. I created a separate: https://github.com/GoogleCloudPlatform/cloud-composer/pull/109

A discussion in PR is the better parts of this work. I love her too. 
 
I work on my English. This is at the top of my personal to-do list. Studies are in a lower position than English lessons. Thank you for understanding.

mik-laj commented on oozie-to-airflow/mappers/mapreduce_mapper.py:
```diff
@@ -0,0 +1,100 @@
+        template_env = jinja2.Environment(loader=template_loader)
+        template = template_env.get_template(self.template)
+        prepare_command = self.get_prepare_command(self.oozie_node, self.params)
+        return template.render(prepare_command=prepare_command, **self.__dict__)
```
Can you use the render_template function here?

sprzedwojski replied:
Good call, it wasn't yet available when I created this mapper.

mik-laj commented on oozie-to-airflow/templates/mapreduce.tpl:
```diff
@@ -0,0 +1,42 @@
+    dataproc_job_id='{{ task_id }}'
+)
+
+{{ task_id }}_prepare.set_downstream({{ task_id }})
```
I do not like defining relation in a template.  In my opinion, the template should only be responsible for displaying relations and not containing logic.

mik-laj replied:
But I do not want you to make changes now.

sprzedwojski replied:
Done, I included the `relations.tpl`

mik-laj commented on oozie-to-airflow/utils/el_utils.py:
```diff
@@ -156,8 +156,20 @@ def parse_els(properties_file: str, prop_dict: Dict[str, str] = None):
+    If the line is a comma-separated list of items, it will be added as a Python list.
+    If it's a single item, it will be added as a string.
+    """
+    props = [x.strip() for x in line.split("=", 1)]
```
I do not want you to make changes, but this loop makes it difficult to understand this code. 
My proposition
```python
    key, value = line.split("=", 1)
    value = replace_el_with_var(value.strip(), prop_dict, quote=False)
    values = value.split(",")
    prop_dict[key.strip()] = values[0] if len(values) <= 1 else values
```

sprzedwojski replied:
This loop was there before, but good idea - your proposition is clearer. I'll change it.

mik-laj commented on oozie-to-airflow/mappers/mapreduce_mapper.py:
```diff
@@ -0,0 +1,100 @@
+        if not output_directory_path:
+            raise Exception("The output_directory_path should be set and is {}".format(output_directory_path))
+
+    def convert_to_airflow_op(self):
```
This method is unnecessary.

sprzedwojski replied:
Removed

mik-laj commented on oozie-to-airflow/mappers/mapreduce_mapper.py:
```diff
@@ -0,0 +1,100 @@
+            self.params_dict = {}
+            for node in param_nodes:
+                param = el_utils.replace_el_with_var(node.text, params=self.params, quote=False)
+                key, value = param.split("=")
```
```suggestion
                key, value = param.split("=", 1)
```
In some cases, this can cause problems.

sprzedwojski replied:
Done

mik-laj commented on oozie-to-airflow/mappers/mapreduce_mapper.py:
```diff
@@ -0,0 +1,100 @@
+
+    def _parse_oozie_node(self):
+        name_node_text = self.oozie_node.find("name-node").text
+        self.name_node = el_utils.replace_el_with_var(name_node_text, params=self.params, quote=False)
```
Where is this field used? I can not find use of this field

mik-laj replied:
I think that you should pass it as a property ``core:fs.defaultFS={self.name_node}``

sprzedwojski replied:
Pass it where? Can you provide any link describing this?

sprzedwojski replied:
After an in-person discussion we've decided to abandon this as it's not currently possible to pass such a claster configuration option the the DAG.

mik-laj replied:
Reference: https://cloud.google.com/dataproc/docs/concepts/configuring-clusters/cluster-properties

sprzedwojski commented on oozie-to-airflow/examples/mapreduce/configuration-template.properties:
```diff
@@ -0,0 +1,38 @@
+# limitations under the License.
+#
+
+dataproc_cluster=CLUSTER-NAME
```
:+1:

sprzedwojski commented on oozie-to-airflow/integration/run_mapreduce_test_with_airflow.bash:
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
```
:+1:

sprzedwojski commented on oozie-to-airflow/integration/prepare_mapreduce_test.bash:
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
```
:+1:

sprzedwojski commented on oozie-to-airflow/mappers/mapreduce_mapper.py:
```diff
@@ -0,0 +1,100 @@
+        return template.render(prepare_command=prepare_command, **self.__dict__)
+
+    def copy_extra_assets(self, input_directory_path: str, output_directory_path: str):
+        self._validate_paths(input_directory_path, output_directory_path)
```
True, copy-pasting ;) I removed it.

potiuk commented on oozie-to-airflow/utils/el_utils.py:
```diff
@@ -162,14 +162,13 @@ def parse_els(properties_file: str, prop_dict: Dict[str, str] = None):
 
 
-def _convert_line(line: str, prop_dict: Dict[str, str]) -> None:
+def _convert_line(line: str, prop_dict: Dict[str, Union[str, list]]) -> None:
```
I don't particularly like using Union of this kind. 

You never know what you get when you do prop_dict[key][0] (first element, or first character). I think it could have been better to always have list[str] as value and in case you need a single value you use prop_dict[key][0]. 

But I also understand that vast majority of cases will be string rather than list, so that's much better to use it this way, though we introduce an implicit coupling and typing does not really work well here.

Another option would be to continue keeping the array value as coma - separated string, which might be actually pretty clean and good solution. If there are just a few properties which are of type array, I find it pretty  ok to do the split(",") after retrieving the property from the dict.

Yet another option - we could have a utility function "get_as_array()" method used in places where you need to retrieve an array. I think that might be by far cleanest method.

We could also monkeypatch dict or add own subclass, but those are not really recommended ways in python.

It's just something to consider, I am ok with either solution eventually.

sprzedwojski replied:
Good call. I've changed it to keep it as a comma-separated string right up till the very end - it's now converted to a list right before being written out to a file.
See: https://github.com/GoogleCloudPlatform/cloud-composer/pull/45/commits/745f1066d125bb219191cc59c721e48581ff870d

potiuk commented on oozie-to-airflow/converter/oozie_converter.py:
```diff
@@ -103,7 +103,7 @@ def _recreate_output_directory(self):
         os.makedirs(self.output_directory_path, exist_ok=True)
 
-    def add_properties_to_params(self, params: Dict[str, str]):
+    def add_properties_to_params(self, params: Dict[str, Any]):
```
I think that's even worse with Any

sprzedwojski replied:
I changed it.
